### PR TITLE
added integration tests - cluster setting page

### PR DESCRIPTION
**Pull request by cyril-ui-developer**
/assign @spadgett

### CODE REVIEW

File: frontend/integration-tests/tests/cluster-settings.scenario.ts [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+    await clusterSettingsView.channelPopupCancelButton.click();
+    await crudView.isLoaded();
+  });
+  it('display Cluster Operators page, click resource item link to display details, click yaml editor, update and save content. Check if updated values persist', async()=>{
+
+    await horizontalnavView.clickHorizontalTab('Cluster Operators');
+    await crudView.isLoaded();
+
+    expect(clusterSettingsView.resourceItemLink.get(4).isDisplayed()).toBe(true);
+    await clusterSettingsView.resourceItemLink.get(4).click();
```
Comment (spadgett) [inline frontend/integration-tests/tests/cluster-settings.scenario.ts]: Are we arbitrarily picking link 4? I suggest have a data-test-id with the component name and clicking the console one.

File: frontend/integration-tests/tests/cluster-settings.scenario.ts [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    expect(clusterSettingsView.channelDropdownButton.isDisplayed()).toBe(true);
+    await clusterSettingsView.channelDropdownButton.click();
+    await crudView.isLoaded();
+
+    expect(clusterSettingsView.getSelectedChannel.get(0).isDisplayed()).toBe(true);
+    const selectedChannelValue = clusterSettingsView.getSelectedChannel.get(0).getText();
+    await clusterSettingsView.getSelectedChannel.get(0).click();
+
+    expect(clusterSettingsView.confirmActionButton.isDisplayed()).toBe(true);
+    await clusterSettingsView.confirmActionButton.click();
```
Comment (spadgett) [inline frontend/integration-tests/tests/cluster-settings.scenario.ts]: A little scary changing the `ClusterVersion` values in our integration tests. We might skip clicking save here. If we do save, we probably want to change it back to the original value.

File: frontend/integration-tests/tests/cluster-settings.scenario.ts [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+  });
+  it('display Cluster Operators page, click resource item link to display details, click yaml editor, update and save content. Check if updated values persist', async()=>{
+
+    await horizontalnavView.clickHorizontalTab('Cluster Operators');
+    await crudView.isLoaded();
+
+    expect(clusterSettingsView.resourceItemLink.get(4).isDisplayed()).toBe(true);
+    await clusterSettingsView.resourceItemLink.get(4).click();
+
+    await horizontalnavView.clickHorizontalTab('YAML');
```
Comment (spadgett) [inline frontend/integration-tests/tests/cluster-settings.scenario.ts]: I would skip the YAML editor tests. Those are pretty well covered by CRUD. Just making sure the link works and details display is probably enough here. We might check something like the page header to know it's the right resource title.

File: frontend/integration-tests/tests/cluster-settings.scenario.ts [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+    expect(clusterSettingsView.viewDetailsButton.get(0).isDisplayed()).toBe(true);
+    await clusterSettingsView.viewDetailsButton.get(0).click();
+    await crudView.isLoaded();
+  });
+  it('display Global Configuration page, click Configuration Resource item link and display details', async() => {
+
+    await horizontalnavView.clickHorizontalTab('Global Configuration');
+    await crudView.isLoaded();
+
+    expect(clusterSettingsView.linkForGlobalConfig.get(4).isPresent()).toBe(true);
```
Comment (spadgett) [inline frontend/integration-tests/tests/cluster-settings.scenario.ts]: I'd suggest again `data-test-id` here instead of always clicking row 4.

File: frontend/integration-tests/tests/cluster-settings.scenario.ts [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    await clusterSettingsView.linkForGlobalConfig.get(4).click();
+    await crudView.isLoaded();
+
+
+    await horizontalnavView.clickHorizontalTab('YAML');
```
Comment (spadgett) [inline frontend/integration-tests/tests/cluster-settings.scenario.ts]: Same comment here... I'd skip modifying the YAML.

File: frontend/integration-tests/tests/cluster-settings.scenario.ts [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+    expect(clusterSettingsView.viewDetailsButton.get(0).isDisplayed()).toBe(true);
+    await clusterSettingsView.viewDetailsButton.get(0).click();
+    await crudView.isLoaded();
+  });
+  it('display Global Configuration page, click dropdown link to edit resource and display details', async() => {
+
+    await horizontalnavView.clickHorizontalTab('Global Configuration');
+    await crudView.isLoaded();
+
+    expect(clusterSettingsView.globalConfigKebabButton.get(0).isDisplayed()).toBe(true);
```
Comment (spadgett) [inline frontend/integration-tests/tests/cluster-settings.scenario.ts]: We shouldn't select the item by index. Let's make sure we select the "Edit" action. Looks like there's already a `data-test-id` you can use.

File: frontend/integration-tests/tests/cluster-settings.scenario.ts [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+    await clusterSettingsView.globalConfigKebabButton.get(0).click();
+    await crudView.isLoaded();
+
+    expect(clusterSettingsView.globalConfigKebabDropDownButton.get(0).isDisplayed()).toBe(true);
+    await clusterSettingsView.globalConfigKebabDropDownButton.get(0).click();
+    await crudView.isLoaded();
+
+    const content = await yamlView.getEditorContent();
+    const newContent = _.defaultsDeep({}, {metadata: {labels}, spec: {text}}, safeLoad(content));
```
Comment (spadgett) [inline frontend/integration-tests/tests/cluster-settings.scenario.ts]: Again I'd skip modifying anything here. Just make sure the YAML editor displays.

File: frontend/integration-tests/tests/cluster-settings.scenario.ts (hunk 6 above)
Comment (spadgett) [reply frontend/integration-tests/tests/cluster-settings.scenario.ts]: I'd add another test checking the Explore API action.

File: frontend/public/components/factory/modal.tsx [hunk 8]
```diff
@@ -82,10 +82,10 @@ export const ModalSubmitFooter: React.SFC<ModalSubmitFooterProps> = ({message, e
 
   return <ModalFooter inProgress={inProgress} errorMessage={errorMessage} message={message}>
     <ActionGroup className="pf-c-form pf-c-form__actions--right pf-c-form__group--no-top-margin">
-      <Button type="button" variant="secondary" onClick={onCancelClick}>{cancelText || 'Cancel'}</Button>
+      <Button type="button" variant="secondary" data-test-id="cancel-action" onClick={onCancelClick}>{cancelText || 'Cancel'}</Button>
       {submitDanger
-        ? <Button type="submit" variant="danger" isDisabled={submitDisabled} id="confirm-action">{submitText}</Button>
-        : <Button type="submit" variant="primary" isDisabled={submitDisabled} id="confirm-action">{submitText}</Button>}
+        ? <Button type="submit" variant="danger" isDisabled={submitDisabled} id="confirm-action" data-test-id="confirm-action">{submitText}</Button>
+        : <Button type="submit" variant="primary" isDisabled={submitDisabled} id="confirm-action" data-test-id="confirm-action">{submitText}</Button>}
```
Comment (spadgett) [inline frontend/public/components/factory/modal.tsx]: We should consider using `data-test-id` instead of `#confirm-action` in other tests. That can be a follow on, though.

File: frontend/integration-tests/tests/cluster-settings.scenario.ts (hunk 1 above)
Comment (cyril-ui-developer) [reply frontend/integration-tests/tests/cluster-settings.scenario.ts]: There is a data-test-id for the resource items but it returns array of links, so I picked index 4. 
I have modified the code and used linkText to select the desired text - console instead of the index. 
Thanks.

File: frontend/public/components/sidebars/resource-sidebar.jsx [hunk 9]
```diff
@@ -34,15 +34,15 @@ export class ResourceSidebarWrapper extends React.Component {
...
+        <button className="btn btn-link" data-test-id="view-schema" onClick={this.toggleSidebar}>
           <InfoCircleIcon className="co-icon-space-r co-p-has-sidebar__sidebar-link-icon" />{linkLabel}
         </button>
       </div>;
     }
 
     return <div className="co-p-has-sidebar__sidebar co-p-has-sidebar__sidebar--bordered hidden-sm hidden-xs" style={{height}}>
       <div className="co-m-pane__body">
-        <button type="button" className="close" aria-label="Close" onClick={this.toggleSidebar}>
+        <button type="button" className="close" aria-label="Close" data-test-id="close" onClick={this.toggleSidebar}>
```
Comment (spadgett) [inline frontend/public/components/sidebars/resource-sidebar.jsx]: I suggest a more precise value like

```suggestion
        <button type="button" className="close" aria-label="Close" data-test-id="sidebar-close" onClick={this.toggleSidebar}>
```

File: frontend/public/components/utils/headings.tsx [hunk 10]
```diff
@@ -50,7 +50,7 @@ export const PageHeading = connectToModel((props: PageHeadingProps) => {
 
   const logo = isCSV
     ? csvLogo()
-    : <div className="co-m-pane__name co-resource-item">{ kind && <ResourceIcon kind={kind} className="co-m-resource-icon--lg" /> } <span id="resource-title" className="co-resource-item__resource-name">{resourceTitle}</span></div>;
+    : <div className="co-m-pane__name co-resource-item">{ kind && <ResourceIcon kind={kind} className="co-m-resource-icon--lg" /> } <span id="resource-title" className="co-resource-item__resource-name" data-test-id="resource-title">{resourceTitle}</span></div>;
```
Comment (spadgett) [inline frontend/public/components/utils/headings.tsx]: I'd remove `id` and update the existing tests that use `#resource-title` to use the `data-test-id`

File: frontend/integration-tests/views/cluster-settings.view.ts [hunk 11]
```diff
@@ -1,5 +1,19 @@
...
+export const viewSchemaButton = $('[data-test-id="view-schema"]');
+export const globalConfigActionsMenu = $('[data-test-id="actions-menu-button"]');
+export const globalConfigEditModelInput = $('[data-test-id="tags-input"]');
+export const detailsCloseButton = $('[data-test-id="close"]');
+export const globalConfigEditLabels = $('[data-test-action="Edit Labels"]');
+export const globalConfigResourceRow = $('[data-test-action="Console"]').$$('[data-test-id="kebab-button"]');
+export const resourceLink = (name) => $(`[data-test-id=${name}]`);
+export const viewDetailsButton = (name)=> $(`[data-test-id=${name}]`);
+export const globalConfigKebabDropDownButton = (name)=> $(`[data-test-action="${name}"]`);
+export const clusterDetailsHeading = (name) => $(`[data-test-id=${name}]`);
```
Comment (spadgett) [inline frontend/integration-tests/views/cluster-settings.view.ts]: If there are only a handful of each, I'd just have constants for the ones we know about. For instance,

```js
export const globalConfigHeading = $('[data-test-id="global-config-resource-header"]');
```

File: frontend/integration-tests/views/cluster-settings.view.ts [hunk 12]
```diff
@@ -1,5 +1,19 @@
...
+export const channelPopupCancelButton = $('[data-test-id="cancel-action"]');
+export const viewSchemaButton = $('[data-test-id="view-schema"]');
+export const globalConfigActionsMenu = $('[data-test-id="actions-menu-button"]');
+export const globalConfigEditModelInput = $('[data-test-id="tags-input"]');
+export const detailsCloseButton = $('[data-test-id="close"]');
+export const globalConfigEditLabels = $('[data-test-action="Edit Labels"]');
+export const globalConfigResourceRow = $('[data-test-action="Console"]').$$('[data-test-id="kebab-button"]');
+export const resourceLink = (name) => $(`[data-test-id=${name}]`);
+export const viewDetailsButton = (name)=> $(`[data-test-id=${name}]`);
+export const globalConfigKebabDropDownButton = (name)=> $(`[data-test-action="${name}"]`);
```
Comment (spadgett) [inline frontend/integration-tests/views/cluster-settings.view.ts]: Can we use `actionForLabel` from crud view?

File: frontend/public/components/utils/selector-input.jsx [hunk 13]
```diff
@@ -93,6 +93,7 @@ export class SelectorInput extends React.Component {
       spellCheck: 'false',
       value: inputValue,
       id: 'tags-input',
+      'data-test-id': 'tags-input',
```
Comment (spadgett) [inline frontend/public/components/utils/selector-input.jsx]: I'd also remove `id` here and update existing tests that reference it

File: frontend/integration-tests/tests/cluster-settings.scenario.ts [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+    await clusterSettingsView.globalConfigResourceRow.click();
+    await crudView.isLoaded();
+
+    expect(clusterSettingsView.globalConfigKebabDropDownButton('Edit Console Resource').isDisplayed()).toBe(true);
+    await clusterSettingsView.globalConfigKebabDropDownButton('Edit Console Resource').click();
+    await crudView.isLoaded();
+    expect(clusterSettingsView.clusterDetailsHeading('resource-title').isDisplayed()).toBe(true);
+    expect(clusterSettingsView.clusterDetailsHeading('resource-title').getText()).toBe('cluster');
+
+    expect(clusterSettingsView.viewSchemaButton.isDisplayed()).toBe(true);
```
Comment (spadgett) [inline frontend/integration-tests/tests/cluster-settings.scenario.ts]: We should definitely have some tests that cover "View Schema," but I'd suggest moving it into a separate PR. It's not specific to cluster settings.

File: frontend/integration-tests/tests/cluster-settings.scenario.ts [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+    await crudView.isLoaded();
+
+    expect(crudView.actionForLabel('Edit Console Resource').isDisplayed()).toBe(true);
+    await crudView.actionForLabel('Edit Console Resource').click();
+    await crudView.isLoaded();
+
+    expect(clusterSettingsView.clusterResourceDetailsTitle.isDisplayed()).toBe(true);
+    expect(clusterSettingsView.clusterResourceDetailsTitle.getText()).toBe('cluster');
+
+    await horizontalnavView.clickHorizontalTab('Overview');
```
Comment (spadgett) [inline frontend/integration-tests/tests/cluster-settings.scenario.ts]: I would skip the tests from here until the end of this function. They're generic details page actions that are already covered by the existing crud scenario.

Comment (spadgett) [reply frontend/integration-tests/tests/cluster-settings.scenario.ts]: Make sure to rollback other changes like new `data-test-id`s if we remove some checks and they're no longer needed.

Comment (cyril-ui-developer) [reply frontend/integration-tests/tests/cluster-settings.scenario.ts]: Will remove them accordingly. Left them for View Schema integration tests which would be another PR, and thought of making use of them soon.

**Comment by spadgett** [comment]
```
2019/09/09 20:46:35 Container cli in pod e2e-aws-console-olm completed successfully
{"component":"entrypoint","file":"prow/entrypoint/run.go:163","func":"k8s.io/test-infra/prow/entrypoint.Options.ExecuteProcess","level":"error","msg":"Entrypoint received interrupt: terminated","time":"2019-09-09T20:55:37Z"}
{"component":"entrypoint","file":"prow/entrypoint/run.go:240","func":"k8s.io/test-infra/prow/entrypoint.gracefullyTerminate","level":"error","msg":"Process did not exit before 15s grace period","time":"2019-09-09T20:55:52Z"}
{"component":"entrypoint","error":"process aborted","file":"prow/entrypoint/run.go:79","func":"k8s.io/test-infra/prow/entrypoint.Options.Run","level":"error","msg":"Error executing test process","time":"2019-09-09T20:55:52Z"}
```

/retest
